src/auth_proxy.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Observer module for Jenkins to auth_proxy integration."""
+
+import logging
+from typing import List
+
+import ops
+from charms.oathkeeper.v0.auth_proxy import AuthProxyConfig, AuthProxyRequirer
+from charms.traefik_k8s.v2.ingress import IngressPerAppRequirer
+
+import jenkins
+import state
+
+AUTH_PROXY_ALLOWED_ENDPOINTS: List[str] = []
+AUTH_PROXY_HEADERS = ["X-User"]
+
+
+logger = logging.getLogger(__name__)
+
+
+class Observer(ops.Object):
+    """The Jenkins Auth Proxy integration observer."""
+
+    def __init__(self, charm: ops.CharmBase, ingress: IngressPerAppRequirer):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: the parent charm to attach the observer to.
+            ingress: the ingress object from which to extract the necessary settings.
+        """
+        super().__init__(charm, "auth-proxy-observer")
+        self.charm = charm
+        self.ingress = ingress
+
+        self.auth_proxy = AuthProxyRequirer(self.charm)
+
+        self.charm.framework.observe(
+            self.charm.on["auth-proxy"].relation_joined, self._auth_proxy_relation_joined
+        )
+        self.charm.framework.observe(
+            self.charm.on["auth-proxy"].relation_departed, self._auth_proxy_relation_departed
+        )
+
+    def _auth_proxy_relation_joined(self, event: ops.RelationCreatedEvent) -> None:
+        """Configure the auth proxy.
+
+        Args:
+            event: the event triggering the handler.
+        """
+        container = self.charm.unit.get_container(state.JENKINS_SERVICE_NAME)
+        if not jenkins.is_storage_ready(container) or not self.ingress.url:
+            logger.warning("Service not yet ready. Deferring.")
+            event.defer()  # The event needs to be handled after Jenkins has started(pebble ready).
+            return
+
+        auth_proxy_config = AuthProxyConfig(
+            protected_urls=[self.ingress.url],
+            allowed_endpoints=AUTH_PROXY_ALLOWED_ENDPOINTS,
+            headers=AUTH_PROXY_HEADERS,
+        )
+        self.auth_proxy.update_auth_proxy_config(auth_proxy_config=auth_proxy_config)
+        # Jenkins Needs to be set up from scratch: requires config and env vars updates.
+        self.charm.on.jenkins_pebble_ready.emit(container)
```

Comment:
Perhaps, rather than emit an event, could we pull out the implementation of Jenkins pebble ready into a separate function, e.g. on the charm, and call it here?

Reply:
I see some issues with that, mainly that we'd be creating a cyclic dependency between this module and the charm module. I think we need to reassess our practices to discuss scenarios like this one

Reply:
@weiiwang01 I think your comment is unnecessary, the problem here has nothing to do with Observer pattern, we should focus on solutions.

@arturo-seijas @jdkandersson
It's explicitly stated in the charm development best practices that this pattern must not be used (https://juju.is/docs/sdk/styleguide#heading--custom-events) and it has been confirmed in Riga during a meeting that included most charm developer leaders and many charm developers. Main reason is that it leads to non implicit code similar to what the reactive pattern was doing, and Ops philosophy is explicitly to avoid the implicit aspects of Reactive.

I don't see why we refactoring the code to extract the common logic required for both `pebble_ready` event and `auth_proxy` events is not possible.
The logic scoped to the `container` and `jenkins` service within the `pebble_ready` event hook method can be extracted into a separate module named `pebble` whose purpose is to manage the pebble plan (as a consequence all methods and constants linked to pebble layers and pebble plans must be moved to this module).
This module is either a class, and as a consequence takes the jenkins service as dependency
Or this module isn't a class and the jenkins service is passed along (becoming a dependency of the proxy observer).

Everything related to charm status or unit version stays in the hook methods (somehow duplicated code)

Reply:
I the it's relevant. If the observer pattern is not applied here, then there's no need for this discussion because there would be no problems. And, we will need no solutions since there are no problems to solve.

Reply:
I have no grudges against the observer pattern; I just have observations regarding its poor real-world performance and am unable to understand the strong preference on it.

The issue of the "calling pebble_ready hook" you mentioned does not exist elsewhere in the charm but is exclusive to the newly introduced observer. Therefore, the changes are unnecessary if no observers are present.

Also, the workarounds introduced to address problems invented by the observer pattern are far from beneficial. For instance, in `state.py`, it must access the relation data directly instead of using the charm library, due to the restriction of observer. In `pebble.py`, object-oriented programming is not feasible without introducing circular dependencies. (<- please don't ask for another refactor; I will approve the current state as is.)

More serious issues include the accessing relation data in `is_auth_proxy_integrated` in `state.py` before `_on_auth_proxy_relation_joined`, as relation data is set during `_on_auth_proxy_relation_joined`. This is another problem with charm state as well.

Handling the auth proxy relation is just a trival task without the observer. By simply running the same config change procedure during the auth proxy relation join and departure events, and in that procedure switch between with and without the auth proxy and it's done. That way is just more simple, easy, clean, compact, and less prone to error.